jupyter/d2l-java/chapter_mlp/weight-decay.ipynb

```diff
@@ -0,0 +1,784 @@
+    "    public InitParams(){\n",
+    "        NDManager manager = NDManager.newBaseManager();\n",
+    "        w = manager.randomNormal(0, 1.0f, new Shape(num_inputs, 1), DataType.FLOAT32, Device.defaultDevice());\n",
+    "        b = manager.zeros(new Shape(1));\n",
+    "        w.attachGradient();\n",
+    "        b.attachGradient();\n",
+    "    }\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Defining $\\ell_2$ Norm Penalty\n",
+    "\n",
+    "Perhaps the most convenient way to implement this penalty\n",
+    "is to square all terms in place and sum them up.\n",
+    "We divide by $2$ by convention\n",
+    "(when we take the derivative of a quadratic function,\n",
+    "the $2$ and $1/2$ cancel out, ensuring that the expression\n",
+    "for the update looks nice and simple)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {
+    "attributes": {
+     "classes": [],
+     "id": "",
+     "n": "3"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "public NDArray l2Penalty(NDArray w){\n",
+    "    return ((w.pow(2)).sum()).div(2);\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "Loss l2loss = Loss.l2Loss();"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Defining the Train and Test Functions\n",
+    "\n",
+    "The following code fits a model on the training set\n",
+    "and evaluates it on the test set.\n",
+    "The linear network and the squared loss\n",
+    "have not changed since the previous chapter,\n",
+    "so we will just import them via `Training.linreg()` and `Training.squaredLoss()`.\n",
+    "The only change here is that our loss now includes the penalty term."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {
+    "attributes": {
+     "classes": [],
+     "id": "",
+     "n": "4"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "float[] trainLoss;\n",
+    "float[] testLoss;\n",
+    "float[] epochCount;\n",
+    "\n",
+    "public void train(float lambd){\n",
+    "    \n",
+    "    InitParams initParams = new InitParams();\n",
+    "    \n",
+    "    NDList params = new NDList(initParams.getW(), initParams.getB());\n",
+    "    \n",
+    "    int num_epochs = 100;\n",
+    "    float lr = 0.003f;\n",
+    "    \n",
+    "    trainLoss = new float[(num_epochs/5)];\n",
+    "    testLoss = new float[(num_epochs/5)];\n",
+    "    epochCount = new float[(num_epochs/5)];\n",
+    "\n",
+    "    for(int epoch = 1; epoch <= num_epochs; epoch++){\n",
+    "        \n",
+    "        for(Batch batch : train_iter.getData(manager)){\n",
+    "            \n",
+    "            NDArray X = batch.getData().head();\n",
+    "            NDArray y = batch.getLabels().head();\n",
+    "            \n",
+    "            // Attach Gradients\n",
```

Comment:
Try to have the comment indentation match the code indentation. Otherwise, it is hard to read